aten/src/ATen/native/mps/operations/RnnOps.mm

```diff
@@ -523,14 +536,22 @@
                 weights.push_back(grad_bias);
                 weights.push_back(grad_bias);
             }
-            gradOutPlaceholder = Placeholder([gradOutputArray objectAtIndex:i], output);
-            gradRecWeightsPlaceholder = Placeholder([gradRecWeightsArray objectAtIndex:i], grad_rec_weights);
-            gradWeightsPlaceholder = Placeholder([gradWeightsArray objectAtIndex:i], grad_weights);
-            gradBiasPlaceholder = Placeholder([gradBiasArray objectAtIndex:i], grad_bias);
-            gradStatePlaceholder = Placeholder([gradStateArray objectAtIndex:i], grad_state);
-            gradCellStatePlaceholder = Placeholder([gradCellStateArray objectAtIndex:i], grad_cell_state);
-
-            [results setObject:gradOutPlaceholder.getMPSGraphTensorData() forKey:gradOutPlaceholder.getMPSGraphTensor()];
+            if (i == 0) {
+                outputPlaceholder = Placeholder([gradOutputArray objectAtIndex:num_layers - i - 1], output_out);
+                gradStatePlaceholder = Placeholder([gradStateArray objectAtIndex:num_layers - i - 1], grad_state_out);
+                gradCellStatePlaceholder = Placeholder([gradCellStateArray objectAtIndex:num_layers - i - 1], grad_cell_state_out);
+            } else {
+                outputPlaceholder = Placeholder([gradOutputArray objectAtIndex:num_layers - i - 1], output);
+                gradStatePlaceholder = Placeholder([gradStateArray objectAtIndex:num_layers - i - 1], grad_state);
+                gradCellStatePlaceholder = Placeholder([gradCellStateArray objectAtIndex:num_layers - i - 1], grad_cell_state);
+            }
+
+            gradRecWeightsPlaceholder = Placeholder([gradRecWeightsArray objectAtIndex:num_layers - i - 1], grad_rec_weights);
+            gradWeightsPlaceholder = Placeholder([gradWeightsArray objectAtIndex:num_layers - i - 1], grad_weights);
+            gradBiasPlaceholder = Placeholder([gradBiasArray objectAtIndex:num_layers - i - 1], grad_bias);
```

Comment:
Notice indices! Elements are stored in the backward order cause they are pushed backwards